### DISCUSSION TITLE
source_to_target / target_to_source message passing

### DISCUSSION THREAD

**Discussion by bryceForrest** [pyg-team/pytorch_geometric#9565] [Q&A]
Hi there! I must be confused.
I am trying to emulate [this](http://staff.ustc.edu.cn/~hexn/papers/sigir20-MBGCN.pdf) model, and ran into some trouble with the user->item and item->user message passing layers. When I used target to source for my item->user message passing layer, none of the parameters were training. Always 0 gradient. However, when I created reverse edges using "ToUndirected", it solved the problem. My intuition tells me these two approaches should be the same, but that's not been my experience.

Minimal example (I'm not calculating the behavior weight or alpha from the paper here because I was trying to make things as barebones as possible for debugging):
```python
class ItemToUser(MessagePassing):
    def __init__(
            self,
            hidden_channels,
            out_channels,
            metadata
        ):
        super().__init__(aggr='mean', flow='target_to_source')
        self.hidden_channels = hidden_channels
        self.out_channels = out_channels
        self.metadata = metadata
        self.linear = Linear(hidden_channels, out_channels, bias=False)

    def forward(self, x_dict, edge_index_dict):
        out = []
        degree_mat = []

        for edge_type, edge_index in edge_index_dict.items():
            src_type, _, dst_type = edge_type
            x_src = x_dict[src_type]
            x_dst = x_dict[dst_type]
            size = x_src.size(0), x_dst.size(0)
            out.append(self.propagate(edge_index, x=x_dst, size=size))
        
        out = torch.stack(out, dim=0).sum(dim=0)
        out = self.linear(out)
        
        return out + x_src

    
class UserToItem(MessagePassing):
    def __init__(
            self,
            hidden_channels,
            out_channels,
            metadata
        ):
        super().__init__(aggr='mean', flow='source_to_target')
        self.hidden_channels = hidden_channels
        self.out_channels = out_channels
        self.metadata = metadata
        self.linear = Linear(hidden_channels, out_channels, bias=False)

    def forward(self, x_dict, edge_index_dict):
        out = []

        for edge_type, edge_index in edge_index_dict.items():
            src_type, _, dst_type = edge_type
            x_src = x_dict[src_type]
            x_dst = x_dict[dst_type]
            size = (x_src.size(0), x_dst.size(0))
            out.append(self.propagate(edge_index, x=x_src, size=size))
        
        out = torch.stack(out, dim=0).sum(dim=0)
        out = self.linear(out)
        
        return out + x_dst
```
The graph is HeteroData with user nodes and item nodes, different edge types representing different behaviors, all edge types going from user to item.
The gradient for ItemToUser's linear layer was always 0, but not UserToItem's. When I added the undirected edges, and put the regular edges through one instance of UserToItem (*really* sending messages from users to items), and the rev edges through another instance of UserToItem (*really* sending messages from items to users), the linear layers for both were training.
What am I missing?!
